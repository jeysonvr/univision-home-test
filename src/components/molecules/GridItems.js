import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { setScrollSlider } from '../../actions/search';
import { imgRequired } from '../../helpers/extra';
import { useFetchGifs } from '../../hooks/useFetchGifs';
import { Button } from '../atoms/Button';
import { GridItem } from '../atoms/GridItem';

export const GridItems = ({ handleRequiredImage }) => {
    
    const divRef = useRef();
    const dispatch = useDispatch();
    const { keyword, offset, scrollSlider } = useSelector(state => state.search);
    const { data:images, loading } = useFetchGifs(keyword, offset);

    const handleScrollLeft = () => {
        divRef.current.scrollLeft -= 160;
    }

    const handleScrollRight = () => {
        divRef.current.scrollLeft += 160;
    }

    const handleScroll = () => {
        dispatch( setScrollSlider(divRef.current.scrollLeft) );

        if( imgRequired(divRef.current) ) {
            handleRequiredImage();
        }
    }

    useEffect(() => {
        if( divRef.current && scrollSlider === 0 ){
            divRef.current.classList.add('fast');
            divRef.current.scrollLeft = 0;
            divRef.current.classList.remove('fast');
        }
    }, [scrollSlider])

    return (
        <>
            { loading && <p className="card">Loading</p> }

            {
                (images.length) 
                ? 
                    <div 
                        className="gridItem__card-grid" 
                        onScroll={ handleScroll } 
                        ref={ divRef }
                    >
                        <div className="gridItem__button left">
                            <Button
                                icon={ 'fas fa-angle-left' }
                                handleClick={ handleScrollLeft }
                            />
                        </div>  
                        <div className="gridItem__button right">
                            <Button
                                handleClick={ handleScrollRight }
                                icon={ 'fas fa-angle-right' }
                            />
                        </div>  
                        <div className="gridMain">
                            { 
                                images.map( img => (
                                    <GridItem 
                                        key={ img.id } 
                                        parentScrollPosition = { scrollSlider }
                                        {...img}
                                    />
                                ))
                            }
                        </div>
                    </div>  
                :
                    <div className="alert alert-primary m-5" role="alert">
                        Search something ...
                    </div>
            }
        </>
    )
}
