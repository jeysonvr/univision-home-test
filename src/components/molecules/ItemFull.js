import React from 'react';
import { useSelector } from 'react-redux';
import { Button } from '../atoms/Button';

export const ItemFull = ({handleCloseModal}) => {

    const { imgURL, imgTitle } = useSelector(state => state.search);

    return (
        <>
            <div className="modal animate__animated animate__fadeIn" tabIndex="-1" style={{display: 'block'}}>
                <div className="modal-dialog modal-dialog-centered">
                    <div className="modal-content text-center">
                        <div className="modal-header">
                            <h5 className="modal-title">{ imgTitle || "No name" }</h5>
                            <Button
                                classList={ 'btn button__close' }
                                handleClick={ handleCloseModal }
                                icon={ 'fas fa-times-circle fa-2x' }
                            />
                        </div>
                        <div className="modal-body card">
                            <img 
                                className="img-fluid"
                                src={ imgURL }
                                alt={ imgTitle }
                            />
                        </div>
                    </div>
                </div>
            </div>
            <div className="modal-backdrop fade show"></div>
        </>
    )
}
