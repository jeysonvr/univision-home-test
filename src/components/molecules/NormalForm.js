import React from 'react';
import { useForm } from '../../hooks/useForm';
import { Button } from '../atoms/Button';
import { Input } from '../atoms/Input';

export const NormalForm = ({ handleSearch }) => {

    const [values, handleInputChange] = useForm({
        keyword: ''
    });
    const { keyword } = values;

    const handleSubmit = (e) => {
        e.preventDefault();
        handleSearch(keyword);
    }

    return (
        <>
            <form onSubmit={ handleSubmit } className="row m-2">
                    <label className="col-2 col-sm-10 col-md-2">Keyword: </label>
                <Input
                    keyword={ keyword }
                    handleInputChange={ handleInputChange }
                />
                <Button 
                    handleClick={ handleSubmit }
                    icon={ '' } 
                    content={'Search'}
                    classList={ 'btn btn-primary col-sm-10 col-md-3 mt-1' }
                />
            </form>
        </>
    )
}
