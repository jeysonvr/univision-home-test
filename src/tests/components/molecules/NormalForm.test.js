import React from 'react';
import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';

import { NormalForm } from '../../../components/molecules/NormalForm';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
    search: {
        keyword: 'test', 
        offset: 0, 
        imgURL: 'a', 
        imgTitle: 'b'
    }
};
let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <NormalForm />
    </Provider>
);

describe('Test in <NormalForm />', () => {

    beforeEach( () => {
        store = mockStore(initialState);
        jest.clearAllMocks();
    })

    test('should show component correctly', () => {
        expect( wrapper ).toMatchSnapshot(); 
    })

})