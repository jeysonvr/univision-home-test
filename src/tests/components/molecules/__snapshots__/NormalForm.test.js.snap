// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <NormalForm /> should show component correctly 1`] = `
<form
  className="row m-2"
  onSubmit={[Function]}
>
  <label
    className="col-2 col-sm-10 col-md-2"
  >
    Keyword: 
  </label>
  <input
    autoComplete="off"
    className="col-12 col-sm-10 col-md-6"
    name="keyword"
    onChange={[Function]}
    placeholder="Search ..."
    style={
      Object {
        "marginRight": "10px",
      }
    }
    type="text"
    value=""
  />
  <button
    className="btn btn-primary col-sm-10 col-md-3 mt-1"
    onClick={[Function]}
  >
    <i
      className=""
    />
    Search
  </button>
</form>
`;
